### PR TITLE
mgr/dashboard: check for existence of Grafana dashboard

### PR DESCRIPTION
**Pull request by a2batic**
If the corresponding Grafana dashboard is not present then the following message will be displayed:-

screenshot from 2018-11-20 12-36-10

Fixes: http://tracker.ceph.com/issues/36356
Signed-off-by: Kanika Murarka <kmurarka@redhat.com>

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html [hunk 1]
```diff
@@ -5,10 +5,17 @@
 <cd-info-panel *ngIf="!grafanaExist"
                i18n>Please consult the
   <a href="{{ docsUrl }}" target="_blank">documentation</a> on how to
-  configure and enable the monitoring functionality.</cd-info-panel>
+  configure and enable the monitoring functionality.
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html]: Please close the element in the same line as the string.
```suggestion
  configure and enable the monitoring functionality.</cd-info-panel>
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html [hunk 2]
```diff
@@ -5,10 +5,17 @@
...
+</cd-info-panel>
+
+<cd-info-panel *ngIf="!dashboardExist"
+               i18n>Dashboard does exist. Please refer to
+  <a href="{{ docsUrl }}" target="_blank">documentation</a> on how to
+  add dashboards to Grafana.
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html]: Please close the element in the same line as the string.
```suggestion
  add dashboards to Grafana.</cd-info-panel>
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/grafana.service.ts [hunk 3]
```diff
@@ -12,4 +12,8 @@ export class GrafanaService {
   getGrafanaApiUrl() {
     return this.http.get('api/grafana/url');
   }
+
+  validateDashboardUrl(uid) {
+    return this.http.get('api/grafana/validation/' + `${uid}`);
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/grafana.service.ts]:
```suggestion
    return this.http.get(`api/grafana/validation/${uid}`);
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html [hunk 4]
```diff
@@ -7,8 +7,13 @@
   <a href="{{ docsUrl }}" target="_blank">documentation</a> on how to
   configure and enable the monitoring functionality.</cd-info-panel>
 
+<cd-info-panel *ngIf="!dashboardExist"
+               i18n>Dashboard does exist. Please refer to
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html]:
```suggestion
               i18n>Dashboard doesn't exist. Please refer to
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/grafana.service.ts (hunk 3 above)
Comment (a2batic) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/grafana.service.ts]: ack

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html (hunk 4 above)
Comment (a2batic) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html]: ack

**Comment by a2batic** [comment]
jenkins retest this please

**Comment by a2batic** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html [hunk 5]
```diff
@@ -7,8 +7,13 @@
   <a href="{{ docsUrl }}" target="_blank">documentation</a> on how to
   configure and enable the monitoring functionality.</cd-info-panel>
 
+<cd-info-panel *ngIf="!dashboardExist"
+               i18n>Dashboard doesn't exist. Please refer to
```
[the remark this answers is not in the document]
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html]: I prefer the current version.
Maybe we add Grafana to differentiate from the dashboard itself:
```suggestion
               i18n>Grafana dashboard doesn't exist. Please refer to
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts [hunk 6]
```diff
@@ -73,7 +78,14 @@ export class GrafanaComponent implements OnInit, OnChanges {
   }
 
   getFrame() {
-    this.url = this.baseUrl + this.grafanaPath + '&refresh=2s' + this.mode;
+    this.settingsService.validateGrafanaDashboardUrl(this.uid).subscribe((data: any) => {
+      if (data !== 200) {
+        this.dashboardExist = false;
```
[the remark this answers is not in the document]
Comment (a2batic) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts]: ack

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (LenzGr) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.html]: > How about
> Suggested change Beta
> [Give us feedback](https://www.research.net/r/SuggestedChanges)
>                i18n>Dashboard doesn't exist. Please refer to
>                i18n>Grafana exists with out Dashboard. Please refer to

That sounds wrong to me, I'd be in favor of @tspmelo 's suggestion.

File: src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf [hunk 7]
```diff
@@ -240,6 +240,14 @@
           <context context-type="sourcefile">app/shared/components/grafana/grafana.component.html</context>
           <context context-type="linenumber">6</context>
         </context-group>
+      </trans-unit><trans-unit id="cbeb993fba46a051afb22a71c5ef10e4681a5280" datatype="html">
+        <source>Dashboard doesn&apos;t exist. Please refer to
```
Comment (LenzGr) [inline src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf]: Dito here:
```suggestion
        <source>Grafana dashboard doesn&apos;t exist. Please refer to
```

Comment (a2batic) [reply src/pybind/mgr/dashboard/frontend/src/locale/messages.xlf]: ack

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts [hunk 8]
```diff
@@ -73,7 +78,14 @@ export class GrafanaComponent implements OnInit, OnChanges {
   }
 
   getFrame() {
-    this.url = this.baseUrl + this.grafanaPath + '&refresh=2s' + this.mode;
+    this.settingsService.validateGrafanaDashboardUrl(this.uid).subscribe((data: any) => {
```
[the remark this answers is not in the document]
Comment (a2batic) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts]: ack

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts [hunk 9]
```diff
@@ -27,11 +27,16 @@ export class GrafanaComponent implements OnInit, OnChanges {
...
+  dashboardExist = true;
 
   @Input()
   grafanaPath: string;
   @Input()
   grafanaStyle: string;
+  @Input()
+  uid: string;
+
+  grafanaUrl: any;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts]: Please remove unused variable "grafanaUrl" as well as "dashboardPath".

Comment (a2batic) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts]: ack

File: src/pybind/mgr/dashboard/tests/test_grafana.py [hunk 10]
```diff
@@ -1,51 +1,26 @@
...
+        self.assertJsonBody({'instance': 'http://localhost:3000'})
 
-        # Test the proxy
-        self._get('/api/grafana/proxy/bar')
-        self.assertStatus(200)
-        self.assertBody('Static Content at path bar')
+    def test_validation(self):
+        self._get('/api/grafana/validation/foo')
+        self.assertStatus(500)
+        # self.assertStatus(200)
```
Comment (p-na) [inline src/pybind/mgr/dashboard/tests/test_grafana.py]: Superfluous?

**Pull request merged**